src/VisualStudio/Core/Def/Implementation/LanguageClient/LiveShareLanguageServerClient.cs

```diff
@@ -0,0 +1,103 @@
+using Microsoft.VisualStudio.Utilities;
+using Nerdbank.Streams;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.LanguageService
+{
+    [ContentType("C#_LSP")]
+    [ContentType("VB_LSP")]
+    [Export(typeof(ILanguageClient))]
+    internal class LiveShareLanguageServerClient : ILanguageClient
+    {
+        private readonly IDiagnosticService _diagnosticService;
+        private readonly IThreadingContext _threadingContext;
+        private readonly LanguageServerProtocol _languageServerProtocol;
+        private readonly Workspace _workspace;
+
+        /// <summary>
+        /// Gets the name of the language client (displayed to the user).
+        /// </summary>
+        public string Name => ServicesVSResources.LiveShare_CSharp_Visual_Basic_Language_Server_Client;
+
+        /// <summary>
+        /// Unused, implementing <see cref="ILanguageClient"/>
+        /// </summary>
+        public IEnumerable<string>? ConfigurationSections { get; } = null;
+
+        /// <summary>
+        /// Gets the initialization options object the client wants to send when 'initialize' message is sent.
+        /// This may be null if the client does not need custom initialization options.
+        /// </summary>
+        public object? InitializationOptions { get; } = null;
+
+        /// <summary>
+        /// Unused, implementing <see cref="ILanguageClient"/>
+        /// </summary>
+        public IEnumerable<string>? FilesToWatch { get; } = null;
+
+        public event AsyncEventHandler<EventArgs>? StartAsync;
+
+        /// <summary>
+        /// Unused, implementing <see cref="ILanguageClient"/>
+        /// </summary>
+        public event AsyncEventHandler<EventArgs>? StopAsync { add { } remove { } }
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public LiveShareLanguageServerClient(LanguageServerProtocol languageServerProtocol, VisualStudioWorkspace workspace, IDiagnosticService diagnosticService, IThreadingContext threadingContext)
+        {
+            _languageServerProtocol = languageServerProtocol;
+            _workspace = workspace;
+            _diagnosticService = diagnosticService;
+            _threadingContext = threadingContext;
+        }
+
+        public Task<Connection> ActivateAsync(CancellationToken token)
+        {
+            var (clientStream, serverStream) = FullDuplexStream.CreatePair();
+            _ = new InProcLanguageServer(serverStream, serverStream, _languageServerProtocol, _workspace, _diagnosticService, _threadingContext);
+            return Task.FromResult(new Connection(clientStream, clientStream));
+        }
+
+        /// <summary>
+        /// Signals that the extension has been loaded.  The server can be started immediately, or wait for user action to start.  
+        /// To start the server, invoke the <see cref="StartAsync"/> event;
+        /// </summary>
+        public async Task OnLoadedAsync()
```

Comment:
So there's a public method that raises an event? This seems fishy.

Reply:
if you have feedback on the LSP client API, we should raise it with them .  I will be filing related feedback for them and will link you offline in case you have other suggestions.